pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,153 @@
+				if compute.Platform.AWS != nil {
+					allErrs = append(allErrs, validateMachinePool(ctx, meta, fldPath.Child("platform", "aws"), config.Platform.AWS, compute.Platform.AWS)...)
+				}
+			}
+		}
+	}
+	return allErrs.ToAggregate()
+}
+
+func validatePlatform(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(platform.Subnets) > 0 {
+		allErrs = append(allErrs, validateSubnets(ctx, meta, fldPath.Child("subnets"), platform.Subnets, networking, publish)...)
+	}
+	return allErrs
+}
+
+func validateSubnets(ctx context.Context, meta *Metadata, fldPath *field.Path, subnets []string, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	privateSubnets, err := meta.PrivateSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	privateSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := privateSubnets[id]; ok {
+			privateSubnetsIdx[id] = idx
+		}
+	}
+	if len(privateSubnets) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, "No private subnets found"))
+	}
+
+	publicSubnets, err := meta.PublicSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	publicSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := publicSubnets[id]; ok {
+			publicSubnetsIdx[id] = idx
+		}
+	}
+
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, privateSubnets, privateSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, publicSubnets, publicSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish != types.InternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
+		errMsg := fmt.Sprintf("No public subnet provided for zones %s", privateZones.Difference(publicZones).List())
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, errMsg))
+	}
+
+	return allErrs
+}
+
+func validateMachinePool(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, pool *awstypes.MachinePool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(pool.Zones) > 0 && len(platform.Subnets) > 0 {
+		privateSubnets, err := meta.PrivateSubnets(ctx)
+		if err != nil {
+			return append(allErrs, field.Invalid(fldPath, meta.subnets, err.Error()))
```

Comment:
You want the subnet field here, right?  Can we just return `err` here?  I expect you would have hit any subnet-retrieval error in `validateSubnets`, which is called by `validatePlatform` before it can call `validateMachinePool`.  Also, seems like we should have a way to say "you know, the subnets are invalid, I'm just going to skip the subnet-dependent part of machine-pool validation".  Not sure how we'd structure that though...

Reply:
Ooh, I like [`field.InternalError`][1] for reporting this `PrivateSubnets` error here.

[1]: https://godoc.org/k8s.io/apimachinery/pkg/util/validation/field#InternalError

Reply:
+1

Reply:
@wking fixed in https://github.com/openshift/installer/compare/a96182f3f4d7c7b69bd674bd918dd72a532f6ce4..9ce0bb39f5cbc2d581705794c8e16933cae350d4